In [1]:
import numpy as np

In [2]:
arr1 = np.array([[1,2,3], [4,5,6], [7,8,9]])

In [3]:
arr1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [4]:
arr2 = arr1.T

In [5]:
arr2

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

# element wise addition

In [6]:
arr1 + arr2

array([[ 2,  6, 10],
       [ 6, 10, 14],
       [10, 14, 18]])

# element wise multiplication

In [7]:
arr1 * arr2

array([[ 1,  8, 21],
       [ 8, 25, 48],
       [21, 48, 81]])

whereas in list, if we do + operation means the lists get merged like:

In [8]:
[324, 452] + [13, 62]

[324, 452, 13, 62]

# element was square root

In [9]:
np.sqrt(arr1)

array([[1.        , 1.41421356, 1.73205081],
       [2.        , 2.23606798, 2.44948974],
       [2.64575131, 2.82842712, 3.        ]])

In [10]:
np.sqrt(arr2)

array([[1.        , 2.        , 2.64575131],
       [1.41421356, 2.23606798, 2.82842712],
       [1.73205081, 2.44948974, 3.        ]])

In [12]:
np.sqrt(arr1+arr2)

array([[1.41421356, 2.44948974, 3.16227766],
       [2.44948974, 3.16227766, 3.74165739],
       [3.16227766, 3.74165739, 4.24264069]])

In [11]:
np.sqrt(arr1*arr2)

array([[1.        , 2.82842712, 4.58257569],
       [2.82842712, 5.        , 6.92820323],
       [4.58257569, 6.92820323, 9.        ]])

In [13]:
arr1.sum()

45

In [14]:
arr2.sum()

45

In [15]:
(arr1+arr2).sum()

90

In [16]:
(arr1*arr2).sum()

261

# searching elements with condition

**where**


In [17]:
arr1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [19]:
np.where(arr1>3)

(array([1, 1, 1, 2, 2, 2], dtype=int64),
 array([0, 1, 2, 0, 1, 2], dtype=int64))

In [20]:
type(np.where(arr1>3))

tuple

np.where(condition) returns a tuple of element indices which satisfy the given condition..

like here array [1,1,1,2,2,2] means [1,1] = element 4>3
[1,2] = element 5>3 and so on

In [21]:
np.where(arr1>5)

(array([1, 2, 2, 2], dtype=int64), array([2, 0, 1, 2], dtype=int64))

**count_nonzero()**

In [23]:
np.count_nonzero(arr1)

9

returns number of elements that are non zeros

In [25]:
np.nonzero(arr1)

(array([0, 0, 0, 1, 1, 1, 2, 2, 2], dtype=int64),
 array([0, 1, 2, 0, 1, 2, 0, 1, 2], dtype=int64))

returns indices where elements are non zero for each axis

In [27]:
arr1[0,2]=0

In [28]:
np.nonzero(arr1)

(array([0, 0, 1, 1, 1, 2, 2, 2], dtype=int64),
 array([0, 1, 0, 1, 2, 0, 1, 2], dtype=int64))

# Python array and Numpy array size comparison

In [29]:
import sys

In [31]:
python_arr = [3,6,2,8,21]

In [32]:
numpy_arr = np.array(python_arr)

In [33]:
sys.getsizeof(1) * len(python_arr)

140

In [34]:
numpy_arr.itemsize * numpy_arr.size

20

**whoa!!!! one element of python array takes 140 bytes of space whereas the same numpy array element takes only 20 bytes of space!!**